Yoast/wordpress-seo
# P1-6 Save schema tab data

## Context


* Fixes [P1-6](https://yoast.atlassian.net/browse/P1-6), [P1-4](https://yoast.atlassian.net/browse/P1-4) and [P1-86](https://yoast.atlassian.net/browse/P1-86).
* There are site-wide schema type options, set in the search appearance. They have defaults too, which depend on the post type (basically posts have article type of `article` and the rest has `none`). You can override these defaults on a per post basis.

## Summary


This PR can be summarized in the following changelog entry:

* Saves the schema tab data to the postmeta and indexables.
* Implements the default option in the schema tab.
* Moves the schema type options to PHP.
* Refactors the schema tab's "fieldgroup" to use the FieldGroup component.

## Relevant technical choices:

* Introduces a default on the hidden input needed to set the default option.
* Override the meta field defaults getting to an empty string for the `schema_page_type` and `schema_article_type`. This is to remove the options from the database so it falls back to the site wide option.
* Changed the select's name property to not be required in https://github.com/Yoast/javascript/pull/809 -- to have them not picked up by the form submit. As they are already represented by the hidden fields.
* Removed the indexable reading that was introduced in https://github.com/Yoast/wordpress-seo/pull/15635 because this PR sets the indexables with the values from the meta options. So the flow now goes options -> indexables, just like the rest.

## Test instructions

This PR can be tested by following these steps:

### Site wide default default
* Go to search appearance settings and set the schema types to the default.
* Create a new post and go to the schema tab.
	* The page type and article type should be set to `Default for Posts` with the default.
* Save the post (you might need to fill in a title and some content to be able to save).
* Check the `wp_postmeta` table for the `_yoast_wpseo_schema_page_type` and `_yoast_wpseo_schema_article_type` that match the `post_id` of the post you just saved. There should be no entries.
  * You can use the query: `SELECT post_id, meta_key, meta_value FROM wp_postmeta WHERE ( meta_key = "_yoast_wpseo_schema_page_type" OR meta_key = "_yoast_wpseo_schema_article_type" ) AND post_id = 335;` and change the `post_id`.
* Check the `wp_yoast_indexable` table's `schema_page_type` and `schema_article_type` fields for that post. They should hold `NULL`.
  * You can use the query: `SELECT object_id, object_type, schema_page_type, schema_article_type FROM wp_yoast_indexable WHERE object_id = 335;` and change the `object_id`.
* Go back to the schema tab of your post and change the values (or one of) to something other than the default.
* Save the post.
* Check the postmeta again. They should hold the values you saved.
* Check the indexable again. The values should hold the values you saved.
* Do the above for pages and custom post types too.

### Validation check
* Go back to the edit page and open the browser console.
* Set the values of the hidden inputs (ids: `yoast_wpseo_schema_page_type` and `yoast_wpseo_schema_article_type`) to something invalid.
  * You can do this by running: `document.getElementById("yoast_wpseo_schema_page_type").value = "blabla"`, switch out the id for the other one.
* Check that the entries in the postmeta table do not exist now.
* Check that the indexable values are `NULL` now.

## UI changes

## Documentation

## Quality assurance


Fixes https://yoast.atlassian.net/browse/P1-6
